### PR TITLE
Fix/navigating find results on html elements

### LINKED ISSUE #14018: Navigating the found keywords in Notebook Find is hard to follow
**Is your feature request related to a problem? Please describe.**
When a "paragraph" finds a number of instances of the found keyword (using Notebook find), it's hard to know which one the arrow / navigation is up to. 

image

**Describe the solution or feature you'd like**
Perhaps the current keyword that is activated by the arrow has a darker color or a different color to make it easier to distinguish.

**Describe alternatives you've considered**
none.

**Additional context**
Found this as a usability issue during Livesite discussion.

### PR DESCRIPTION
**Pull request by MaddyDev**
This PR fixes #14018 
fixFindInMarkdown

### CODE REVIEW

File: src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts [hunk 1]
```diff
@@ -430,8 +424,8 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
 		let textOutput: string[] = [];
 		let elements = this.getHtmlElements();
 		elements.forEach(element => {
-			if (element && element.innerText) {
-				textOutput.push(element.innerText);
+			if (element && element.textContent) {
```
Comment (chlafreniere) [inline src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:427]: can you describe the change from innerText to textContent?

Comment (MaddyDev) [reply src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:427]: text content is what mark js is basing it's ranges for highlight. innerText sometimes is missing some spaces and newlines due to which the range is off. When we use textContent to calculate the ranges, we can directly use that range to highlight the correct range instead of marking all the elements with the keyword.

**Comment by chlafreniere** [comment]
@MaddyDev this would be a great time to create a notebook that everyone can use to test find functionality for releases. We can put it in the *-docs repo.

Some quick ideas for things to put in that notebook:
- Tables
- Links
- Bolded/italicized/underlined items
- Elements with one match
- Elements with more than one match (e.g. a paragraph with the word "test" twice)

**Comment by MaddyDev** [comment]
> @MaddyDev this would be a great time to create a notebook that everyone can use to test find functionality for releases. We can put it in the *-docs repo.
> 
> Some quick ideas for things to put in that notebook:
> 
> * Tables
> * Links
> * Bolded/italicized/underlined items
> * Elements with one match
> * Elements with more than one match (e.g. a paragraph with the word "test" twice)

Sure thing! Will work on creating a notebook for that.

File: src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts [hunk 2]
```diff
@@ -386,6 +379,7 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
 			let elementContainingText = elements[range.startLineNumber - 1];
 			let mark = new Mark(elementContainingText);
 			mark.unmark({ acrossElements: true, className: 'rangeHighlight' });
+			mark.unmark({ acrossElements: true, className: 'rangespecifcHighlight' });
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:382]: is there a reason you're not capitalizing specific? `rangeSpecificHighlight`

**Review by Charles-Gagnon — COMMENTED** [review]
In the linked issue @radentonev mentioned that it would be nice if we had it so that it still highlighted all words (in yellow for example) - but then the "currently selected" word would have its own color (e.g. orange)

It looks like that isn't being done here - is there a reason you went this route instead? It's definitely nice to be able to quickly see all the words on a page that match the search term without having to click through them all.

**Comment by MaddyDev** [comment]
> In the linked issue @radentonev mentioned that it would be nice if we had it so that it still highlighted all words (in yellow for example) - but then the "currently selected" word would have its own color (e.g. orange)
> 
> It looks like that isn't being done here - is there a reason you went this route instead? It's definitely nice to be able to quickly see all the words on a page that match the search term without having to click through them all.

Previously we didn't know of a way to highlight the exact word in the element so we used to highlight the entire element and then started using mark js to highlight all the words that match the keyword. I was able to find a way to highlight the exact word now so I made the above changes. I can actually go back to highlighting all the words and then use orange for current index. Wanted to know the preference, so I asked Raden and PM's on the issue.

File: src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts (hunk 2 above)
Comment (MaddyDev) [reply src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:382]: this was supposed to be removed, but yeah I think I just missed the casing there.

**Comment by radentonev** [comment]
@MaddyDev my preference is to highlight all in yellow and current index in orange. I think this would work best.

**Comment by Charles-Gagnon** [comment]
@radentonev I think we should probably actually just use the same highlight colors that the other editors use so we're consistent across the app - that sound reasonable?

**Comment by radentonev** [comment]
@Charles-Gagnon yeah absolutely. The example I gave with yellow and orange are coming from the Md editor in ADS, so I felt like matching that might make sense.

**Comment by radentonev** [comment]
@Charles-Gagnon I see so the images you pasted are from Query editor. Looks like we use different colors between MD files and SQL files. I am fine with using either one. We can revisit the colors in the future and perhaps standardize, but I am fine with following either one for the time being.

**Comment by Charles-Gagnon** [comment]
Yeah, I agree. @MaddyDev so pick one that looks the best and make sure to re-use the colors defined for the one you choose.

**Comment by MaddyDev** [comment]
@radentonev @MsSQLGirl @Charles-Gagnon  @chlafreniere 
How does the below look: (Agree we can standardize the colors later)
fixFindSpecific

**Comment by radentonev** [comment]
@MaddyDev Iooks great to me.

**Comment by MsSQLGirl** [comment]
@MaddyDev - looks awesome :) Love it!

**Review by corivera — APPROVED** [review]

**Review by lucyzhang929 — APPROVED** [review]

File: src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts [hunk 3]
```diff
@@ -370,22 +370,31 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
...
 							className: 'rangeHighlight'
 						});
 						elementContainingText.scrollIntoView({ behavior: 'smooth' });
 					}
 				}
+				mark.markRanges([{
+					start: range.startColumn - 1, //subtracting 1 since markdown html is 0 indexed.
+					length: range.endColumn - range.startColumn
+				}], {
+					className: 'rangeSpecifcHighlight'
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:383]: typo here - `rangeSpecificHighlight` (make sure to update the other references too)

I'd suggest making this into a const in the file to make this easier and safer.

File: src/sql/workbench/contrib/notebook/browser/cellViews/textCell.css [hunk 4]
```diff
@@ -42,11 +42,19 @@ text-cell-component code-component .monaco-editor .inputarea.ime-input {
...
+	background-color: yellow;
 }
 
 .vs-dark .notebook-preview .rangeHighlight {
-	background-color: rgba(255, 255, 0, 0.2)
+	background-color: yellow;
+}
+
+.vs .notebook-preview .rangeSpecifcHighlight {
+	background-color: orange;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/textCell.css:53]: We shouldn't be hardcoding color values like this. They should really be defined as a themed color so that users can update them in themes as they wish - for example here's the general editor ones : https://github.com/microsoft/azuredatastudio/blob/main/src/vs/platform/theme/common/colorRegistry.ts#L304

I'm not sure if the md colors are registered like this - but if they aren't then we should still make new entries for this to allow for this functionality.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: src/sql/workbench/contrib/notebook/browser/cellViews/textCell.css (hunk 4 above)
Comment (chlafreniere) [reply src/sql/workbench/contrib/notebook/browser/cellViews/textCell.css:53]: As a refresher, we can leverage notebookStyles for something like this.

File: resources/xlf/LocProject.json [hunk 5]
```diff
@@ -1,93 +0,0 @@
-{
-  "Projects": [
```
Comment (Charles-Gagnon) [inline resources/xlf/LocProject.json:2]: undo this delete

File: src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts [hunk 6]
```diff
@@ -361,6 +361,7 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
 
 	private addDecoration(range: NotebookRange): void {
 		if (range && this.output && this.output.nativeElement) {
+			let markDoc = new Mark(this.output.nativeElement);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:364]: I think the name could be better here - `markDoc` isn't very clear. Maybe `markOutput` or `outputMark`?

Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:364]: It might be helpful to explain in a comment why the two different Mark objects are needed

**Comment by Charles-Gagnon** [comment]
One thing I noticed - it seems like the search highlighting is only across a single cell/output. I would think that if someone is searching they want to search across the entire Notebook though. Is this something done on purpose on your part? It looks little odd in your video where you reach the end of the search for the current "section" and then it immediately jumps into a new section with a bunch of other highlights. 

And then to verify - is this search doing both cell content and outputs?

File: resources/xlf/LocProject.json (hunk 5 above)
Comment (MaddyDev) [reply resources/xlf/LocProject.json:2]: i just excluded this, somehow it got deleted. Let me undo it. Thanks

**Comment by MaddyDev** [comment]
> One thing I noticed - it seems like the search highlighting is only across a single cell/output. I would think that if someone is searching they want to search across the entire Notebook though. Is this something done on purpose on your part? It looks little odd in your video where you reach the end of the search for the current "section" and then it immediately jumps into a new section with a bunch of other highlights.
> 
> And then to verify - is this search doing both cell content and outputs?

Yeah it's on purpose, not because we want this experience. But because there's no way to access the entire notebook elements since each cell is an editor and we have to iterate through all the editors to apply this. It's definitely do-able but requires a larger set of changes. Am addressing this specific to the shortcomings we have with the current way.

**Comment by MaddyDev** [comment]
> And then to verify - is this search doing both cell content and outputs?

It only searches the cell contents not the outputs. That's also an ask we have in the backlog.

File: resources/xlf/LocProject.json [hunk 7]
```diff
@@ -1,93 +1,93 @@
...
-    {
-      "LanguageSet": "AzureKatal_Languages",
-      "LocItems": [
-        {
-          "SourceFile": "resources\\xlf\\en\\admin-tool-ext-win.xlf",
-          "LclFile": "resources\\localization\\LCL\\{Lang}\\admin-tool-ext-win.xlf.lcl",
-          "CopyOption": "LangIDOnName",
-          "OutputPath": "resources\\xlf\\Localized"
-        },
+	"Projects": [
```
Comment (Charles-Gagnon) [inline resources/xlf/LocProject.json:2]: It looks like you've modified the file now with the latest changes - make sure you're copying the file exactly as it was

File: src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts [hunk 8]
```diff
@@ -361,31 +361,41 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
 
 	private addDecoration(range: NotebookRange): void {
 		if (range && this.output && this.output.nativeElement) {
+			let markOutput = new Mark(this.output.nativeElement); // to highlight all occurances in the element.
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:364]: Maybe something like `markAllOccurances` instead? I'm not sure what output has to do with this - it definitely confused me since I thought it was referring to actual cell output. 

And then the mark below could be `markSelectedText`

**Review by Charles-Gagnon — APPROVED** [review]
Looks good once you've fixed the issue with the JSON.

Although I did leave a comment about the name of the properties too if you could take a look at that.

File: resources/xlf/LocProject.json (hunk 7 above)
Comment (MaddyDev) [reply resources/xlf/LocProject.json:2]: it doesn't let me commit when I have the file as is, I did format document. Is that ok?

Comment (Charles-Gagnon) [reply resources/xlf/LocProject.json:2]: Please don't modify files (especially ones not used by our team).

You should checkout main and then copy that version of the file back into this branch.

File: resources/xlf/LocProject.json [hunk 9]
```diff
@@ -90,4 +90,4 @@
       ]
     }
   ]
-}
\ No newline at end of file
+}
```
Comment (MaddyDev) [inline resources/xlf/LocProject.json:93]: @Charles-Gagnon I copied content from main branch and saved it, for some reason it adds new line at the end when I save it. Not able to find an alternative, do you know what I can do here?

Comment (Charles-Gagnon) [reply resources/xlf/LocProject.json:93]: Can you not delete the newline?

Comment (Charles-Gagnon) [reply resources/xlf/LocProject.json:93]: Anyways if you can't get it working for some reason it's fine to leave this one I guess - trying to revert it at this point would just be more of a pain than is worth it
